File: packages/java/src/main/java/com/missioninbox/resources/DomainRedirects.java
package com.missioninbox.resources;

import com.missioninbox.http.HttpClient;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Domain redirects sub-resource. Access via {@code mi.domains.redirects}. */
public final class DomainRedirects {

    private final HttpClient http;

    public DomainRedirects(HttpClient http) {
        this.http = http;
    }

    /** Return the IP + CNAME target values for manual DNS publishing. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> getDnsConfig() {
        return (Map<String, Object>) http.request(
                "GET", "/api/domains/redirect/dns-config", null, null);
    }

    /** Get the current redirect configuration for a domain. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> get(String domainName) {
        return (Map<String, Object>) http.request(
                "GET", "/api/domains/" + encode(domainName) + "/redirect", null, null);
    }

    /**
     * Create or update the redirect for a domain.
     *
     * <p>Recognised keys: {@code redirectUrl} (required), {@code enabled}, {@code forceHttps}.
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> setup(String domainName, Map<String, Object> params) {
        return (Map<String, Object>) http.request(
                "PUT", "/api/domains/" + encode(domainName) + "/redirect", null, params);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> pushDns(String domainName) {
        return (Map<String, Object>) http.request(
                "POST", "/api/domains/" + encode(domainName) + "/redirect/push-dns", null, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> verifyDns(String domainName) {
        return (Map<String, Object>) http.request(
                "POST", "/api/domains/" + encode(domainName) + "/redirect/verify-dns", null, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> getEvents(String domainName, Integer limit) {
        Map<String, Object> query = limit == null ? null : Collections.singletonMap("limit", limit);
        return (Map<String, Object>) http.request(
                "GET", "/api/domains/" + encode(domainName) + "/redirect/events", query, null);
    }

    public Map<String, Object> getEvents(String domainName) {
        return getEvents(domainName, null);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> delete(String domainName) {
        return (Map<String, Object>) http.request(
                "DELETE", "/api/domains/" + encode(domainName) + "/redirect", null, null);
    }

    /** Create redirects on many domains. Each item needs {@code domainName}, {@code redirectUrl}. */
    @SuppressWarnings("unchecked")
    public Map<String, Object> bulkSetup(List<Map<String, Object>> redirects) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("redirects", redirects);
        return (Map<String, Object>) http.request(
                "POST", "/api/domains/redirects/bulk-setup", null, body);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> bulkUpdate(List<Map<String, Object>> updates) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("updates", updates);
        return (Map<String, Object>) http.request(
                "PATCH", "/api/domains/redirects/bulk-update", null, body);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> bulkDelete(List<String> domainNames) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("domainNames", domainNames);
        return (Map<String, Object>) http.request(
                "DELETE", "/api/domains/redirects/bulk-delete", null, body);
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> bulkCreateOrUpdate(List<Map<String, Object>> redirects) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("redirects", redirects);
        return (Map<String, Object>) http.request(
                "POST", "/api/domains/redirects/bulk-create-or-update", null, body);
    }

    private static String encode(String s) {
        return URLEncoder.encode(s, StandardCharsets.UTF_8);
    }
}

File: packages/python/src/missioninbox/resources/domain_redirects.py
"""Domain redirects sub-resource. Access via `mi.domains.redirects`."""

from __future__ import annotations

from typing import Any, Optional
from urllib.parse import quote

from .._http import HttpClient


class DomainRedirects:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def get_dns_config(self) -> dict[str, Any]:
        """Return the IP + CNAME target values for manual DNS publishing."""
        return self._http.request("GET", "/api/domains/redirect/dns-config")

    def get(self, domain_name: str) -> dict[str, Any]:
        return self._http.request("GET", f"/api/domains/{quote(domain_name, safe='')}/redirect")

    def setup(
        self,
        domain_name: str,
        *,
        redirect_url: str,
        enabled: Optional[bool] = None,
        force_https: Optional[bool] = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"redirectUrl": redirect_url}
        if enabled is not None:
            body["enabled"] = enabled
        if force_https is not None:
            body["forceHttps"] = force_https
        return self._http.request(
            "PUT",
            f"/api/domains/{quote(domain_name, safe='')}/redirect",
            body=body,
        )

    def push_dns(self, domain_name: str) -> dict[str, Any]:
        return self._http.request(
            "POST",
            f"/api/domains/{quote(domain_name, safe='')}/redirect/push-dns",
        )

    def verify_dns(self, domain_name: str) -> dict[str, Any]:
        return self._http.request(
            "POST",
            f"/api/domains/{quote(domain_name, safe='')}/redirect/verify-dns",
        )

    def get_events(self, domain_name: str, limit: Optional[int] = None) -> dict[str, Any]:
        query = {"limit": limit} if limit is not None else None
        return self._http.request(
            "GET",
            f"/api/domains/{quote(domain_name, safe='')}/redirect/events",
            query=query,
        )

    def delete(self, domain_name: str) -> dict[str, Any]:
        return self._http.request(
            "DELETE",
            f"/api/domains/{quote(domain_name, safe='')}/redirect",
        )

    def bulk_setup(self, *, redirects: list[dict[str, Any]]) -> dict[str, Any]:
        return self._http.request(
            "POST",
            "/api/domains/redirects/bulk-setup",
            body={"redirects": redirects},
        )

    def bulk_update(self, *, updates: list[dict[str, Any]]) -> dict[str, Any]:
        return self._http.request(
            "PATCH",
            "/api/domains/redirects/bulk-update",
            body={"updates": updates},
        )

    def bulk_delete(self, *, domain_names: list[str]) -> dict[str, Any]:
        return self._http.request(
            "DELETE",
            "/api/domains/redirects/bulk-delete",
            body={"domainNames": domain_names},
        )

    def bulk_create_or_update(self, *, redirects: list[dict[str, Any]]) -> dict[str, Any]:
        return self._http.request(
            "POST",
            "/api/domains/redirects/bulk-create-or-update",
            body={"redirects": redirects},
        )
